File: example/c#/CSharpExample/CSharpExample/TestModule.cs
using System.Collections.Generic;
using Newtonsoft.Json;
using System.IO;
using UnityEngine;
using System;



// 测试用例表
public partial class TestConfig
{

	// 字符串
	public string stringfield;

	// bool型
	public bool boolfield;

	// 字符
	public char charfield;

	// 16位整数
	public int int16field;

	// 32位整数
	public int intfield;

	// 64位整数
	public long int64field;

	// 8位无符号整数
	public int uint8field;

	// 16位无符号整数
	public int uint16field;

	// 32位无符号整数
	public int uintfield;

	// 64位无符号整数
	public long uint64field;

	// 浮点数
	public float floatfield;

	// 双精度浮点数
	public double doublefield;

	// 字符串主键
	public string stringfieldprimary;

	// bool型主键
	public bool boolfieldprimary;

	// 字符主键
	public char charfieldprimary;

	// 16位整数主键
	public int int16fieldprimary;

	// 32位整数主键
	public int intfieldprimary;

	// 64位整数主键
	public long int64fieldprimary;

	// 8位无符号整数主键
	public int uint8fieldprimary;

	// 16位无符号整数主键
	public int uint16fieldprimary;

	// 32位无符号整数主键
	public int uintfieldprimary;

	// 64位无符号整数主键
	public long uint64fieldprimary;

	// 浮点数主键
	public float floatfieldprimary;

	// 双精度浮点数主键
	public double doublefieldprimary;

	// 字符串唯一
	public string stringfieldunique;

	// bool型唯一
	public bool boolfieldunique;

	// 字符唯一
	public char charfieldunique;

	// 16位整数唯一
	public int int16fieldunique;

	// 32位整数唯一
	public int intfieldunique;

	// 64位整数唯一
	public long int64fieldunique;

	// 8位无符号整数唯一
	public int uint8fieldunique;

	// 16位无符号整数唯一
	public int uint16fieldunique;

	// 32位无符号整数唯一
	public int uintfieldunique;

	// 64位无符号整数唯一
	public long uint64fieldunique;

	// 浮点数唯一
	public float floatfieldunique;

	// 双精度浮点数唯一
	public double doublefieldunique;

	// 字符串数组
	public string m_liststringfieldarray;

	// bool型数组
	public bool m_listboolfieldarray;

	// 字符数组
	public char m_listcharfieldarray;

	// 16位整数数组
	public int m_listint16fieldarray;

	// 32位整数数组
	public int m_listintfieldarray;

	// 64位整数数组
	public long m_listint64fieldarray;

	// 8位无符号整数数组
	public int m_listuint8fieldarray;

	// 16位无符号整数数组
	public int m_listuint16fieldarray;

	// 32位无符号整数数组
	public int m_listuintfieldarray;

	// 64位无符号整数数组
	public long m_listuint64fieldarray;

	// 浮点数数组
	public float m_listfloatfieldarray;

	// 双精度浮点数数组
	public double m_listdoublefieldarray;

	// 字符串集合
	public string m_setstringfieldset;

	// bool型集合
	public bool m_setboolfieldset;

	// 字符集合
	public char m_setcharfieldset;

	// 16位整数集合
	public int m_setint16fieldset;

	// 32位整数集合
	public int m_setintfieldset;

	// 64位整数集合
	public long m_setint64fieldset;

	// 8位无符号整数集合
	public int m_setuint8fieldset;

	// 16位无符号整数集合
	public int m_setuint16fieldset;

	// 32位无符号整数集合
	public int m_setuintfieldset;

	// 64位无符号整数集合
	public long m_setuint64fieldset;

	// 浮点数集合
	public float m_setfloatfieldset;

	// 双精度浮点数集合
	public double m_setdoublefieldset;
 }


public partial class TestModule
{		
	public static TestModule instance = new TestModule();
	
	
	// 测试用例表
	public List<TestConfig> m_listTestConfig = new List<TestConfig>();

	

	// 字符串唯一
	public Dictionary<string, TestConfig> m_mapTestConfigBystringfieldunique = new Dictionary<string, TestConfig>();
	
	// bool型唯一
	public Dictionary<bool, TestConfig> m_mapTestConfigByboolfieldunique = new Dictionary<bool, TestConfig>();
	
	// 字符唯一
	public Dictionary<char, TestConfig> m_mapTestConfigBycharfieldunique = new Dictionary<char, TestConfig>();
	
	// 16位整数唯一
	public Dictionary<int, TestConfig> m_mapTestConfigByint16fieldunique = new Dictionary<int, TestConfig>();
	
	// 32位整数唯一
	public Dictionary<int, TestConfig> m_mapTestConfigByintfieldunique = new Dictionary<int, TestConfig>();
	
	// 64位整数唯一
	public Dictionary<long, TestConfig> m_mapTestConfigByint64fieldunique = new Dictionary<long, TestConfig>();
	
	// 8位无符号整数唯一
	public Dictionary<int, TestConfig> m_mapTestConfigByuint8fieldunique = new Dictionary<int, TestConfig>();
	
	// 16位无符号整数唯一
	public Dictionary<int, TestConfig> m_mapTestConfigByuint16fieldunique = new Dictionary<int, TestConfig>();
	
	// 32位无符号整数唯一
	public Dictionary<int, TestConfig> m_mapTestConfigByuintfieldunique = new Dictionary<int, TestConfig>();
	
	// 64位无符号整数唯一
	public Dictionary<long, TestConfig> m_mapTestConfigByuint64fieldunique = new Dictionary<long, TestConfig>();
	
	// 浮点数唯一
	public Dictionary<float, TestConfig> m_mapTestConfigByfloatfieldunique = new Dictionary<float, TestConfig>();
	
	// 双精度浮点数唯一
	public Dictionary<double, TestConfig> m_mapTestConfigBydoublefieldunique = new Dictionary<double, TestConfig>();
	

	
	
	// 字符串唯一
	public TestConfig GetTestConfigBystringfieldunique(string stringfieldunique)
	{
		TestConfig ret;
		m_mapTestConfigBystringfieldunique.TryGetValue(stringfieldunique, out ret);
		return ret;
	}
	
	// bool型唯一
	public TestConfig GetTestConfigByboolfieldunique(bool boolfieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByboolfieldunique.TryGetValue(boolfieldunique, out ret);
		return ret;
	}
	
	// 字符唯一
	public TestConfig GetTestConfigBycharfieldunique(char charfieldunique)
	{
		TestConfig ret;
		m_mapTestConfigBycharfieldunique.TryGetValue(charfieldunique, out ret);
		return ret;
	}
	
	// 16位整数唯一
	public TestConfig GetTestConfigByint16fieldunique(int int16fieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByint16fieldunique.TryGetValue(int16fieldunique, out ret);
		return ret;
	}
	
	// 32位整数唯一
	public TestConfig GetTestConfigByintfieldunique(int intfieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByintfieldunique.TryGetValue(intfieldunique, out ret);
		return ret;
	}
	
	// 64位整数唯一
	public TestConfig GetTestConfigByint64fieldunique(long int64fieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByint64fieldunique.TryGetValue(int64fieldunique, out ret);
		return ret;
	}
	
	// 8位无符号整数唯一
	public TestConfig GetTestConfigByuint8fieldunique(int uint8fieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByuint8fieldunique.TryGetValue(uint8fieldunique, out ret);
		return ret;
	}
	
	// 16位无符号整数唯一
	public TestConfig GetTestConfigByuint16fieldunique(int uint16fieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByuint16fieldunique.TryGetValue(uint16fieldunique, out ret);
		return ret;
	}
	
	// 32位无符号整数唯一
	public TestConfig GetTestConfigByuintfieldunique(int uintfieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByuintfieldunique.TryGetValue(uintfieldunique, out ret);
		return ret;
	}
	
	// 64位无符号整数唯一
	public TestConfig GetTestConfigByuint64fieldunique(long uint64fieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByuint64fieldunique.TryGetValue(uint64fieldunique, out ret);
		return ret;
	}
	
	// 浮点数唯一
	public TestConfig GetTestConfigByfloatfieldunique(float floatfieldunique)
	{
		TestConfig ret;
		m_mapTestConfigByfloatfieldunique.TryGetValue(floatfieldunique, out ret);
		return ret;
	}
	
	// 双精度浮点数唯一
	public TestConfig GetTestConfigBydoublefieldunique(double doublefieldunique)
	{
		TestConfig ret;
		m_mapTestConfigBydoublefieldunique.TryGetValue(doublefieldunique, out ret);
		return ret;
	}
	

	
	// 加载单个文件
	public bool LoadTestConfig(string basePath)
	{
		string jsonPath = basePath + @"TestConfig.json";
		string jsonText = File.ReadAllText(jsonPath, System.Text.Encoding.Default);
		
		try
		{
			m_listTestConfig = (List<TestConfig>)Newtonsoft.Json.JsonConvert.DeserializeObject(jsonText, typeof(List<TestConfig>));
		}
		catch (Exception e)
		{
			Debug.Log("加载" + jsonPath + "配置失败，原因 = " + e.Message);
		}			

		for (int i = 0; i < m_listTestConfig.Count; ++i)
		{
			TestConfig row = m_listTestConfig[i];
				m_mapTestConfigBystringfieldunique[row.stringfieldunique] = row;
				m_mapTestConfigByboolfieldunique[row.boolfieldunique] = row;
				m_mapTestConfigBycharfieldunique[row.charfieldunique] = row;
				m_mapTestConfigByint16fieldunique[row.int16fieldunique] = row;
				m_mapTestConfigByintfieldunique[row.intfieldunique] = row;
				m_mapTestConfigByint64fieldunique[row.int64fieldunique] = row;
				m_mapTestConfigByuint8fieldunique[row.uint8fieldunique] = row;
				m_mapTestConfigByuint16fieldunique[row.uint16fieldunique] = row;
				m_mapTestConfigByuintfieldunique[row.uintfieldunique] = row;
				m_mapTestConfigByuint64fieldunique[row.uint64fieldunique] = row;
				m_mapTestConfigByfloatfieldunique[row.floatfieldunique] = row;
				m_mapTestConfigBydoublefieldunique[row.doublefieldunique] = row;
		}
		
		return true;
	}

	
	// 加载配置
	public bool Load(string basePath)
	{
		if (!LoadTestConfig(basePath))
		{
			return false;
		}
		
		return true;
	}

}






	

File: example/c++/c++_example/cfg/TestModule.cpp
//------------------------------------------------------------------------------
// 本头文件由工具自动生成，请勿在本文件内做改动
// 日期：
//------------------------------------------------------------------------------

#include "TestModule.h"
#include <iostream>
#include <rapidjson/document.h>	
#include "tokit_tool.h"




TestModule TestModule::instance;





// 字符串唯一
const TestConfig* TestModule::FindTestConfigBystringfieldunique(std::string stringfieldunique)
{
	const auto itr = m_mapTestConfigBystringfieldunique.find(stringfieldunique);
	if (itr == m_mapTestConfigBystringfieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// bool型唯一
const TestConfig* TestModule::FindTestConfigByboolfieldunique(bool boolfieldunique)
{
	const auto itr = m_mapTestConfigByboolfieldunique.find(boolfieldunique);
	if (itr == m_mapTestConfigByboolfieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 字符唯一
const TestConfig* TestModule::FindTestConfigBycharfieldunique(char charfieldunique)
{
	const auto itr = m_mapTestConfigBycharfieldunique.find(charfieldunique);
	if (itr == m_mapTestConfigBycharfieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 16位整数唯一
const TestConfig* TestModule::FindTestConfigByint16fieldunique(int int16fieldunique)
{
	const auto itr = m_mapTestConfigByint16fieldunique.find(int16fieldunique);
	if (itr == m_mapTestConfigByint16fieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 32位整数唯一
const TestConfig* TestModule::FindTestConfigByintfieldunique(int intfieldunique)
{
	const auto itr = m_mapTestConfigByintfieldunique.find(intfieldunique);
	if (itr == m_mapTestConfigByintfieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 64位整数唯一
const TestConfig* TestModule::FindTestConfigByint64fieldunique(long int64fieldunique)
{
	const auto itr = m_mapTestConfigByint64fieldunique.find(int64fieldunique);
	if (itr == m_mapTestConfigByint64fieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 8位无符号整数唯一
const TestConfig* TestModule::FindTestConfigByuint8fieldunique(int uint8fieldunique)
{
	const auto itr = m_mapTestConfigByuint8fieldunique.find(uint8fieldunique);
	if (itr == m_mapTestConfigByuint8fieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 16位无符号整数唯一
const TestConfig* TestModule::FindTestConfigByuint16fieldunique(int uint16fieldunique)
{
	const auto itr = m_mapTestConfigByuint16fieldunique.find(uint16fieldunique);
	if (itr == m_mapTestConfigByuint16fieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 32位无符号整数唯一
const TestConfig* TestModule::FindTestConfigByuintfieldunique(int uintfieldunique)
{
	const auto itr = m_mapTestConfigByuintfieldunique.find(uintfieldunique);
	if (itr == m_mapTestConfigByuintfieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 64位无符号整数唯一
const TestConfig* TestModule::FindTestConfigByuint64fieldunique(int64_t uint64fieldunique)
{
	const auto itr = m_mapTestConfigByuint64fieldunique.find(uint64fieldunique);
	if (itr == m_mapTestConfigByuint64fieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 浮点数唯一
const TestConfig* TestModule::FindTestConfigByfloatfieldunique(float floatfieldunique)
{
	const auto itr = m_mapTestConfigByfloatfieldunique.find(floatfieldunique);
	if (itr == m_mapTestConfigByfloatfieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}

// 双精度浮点数唯一
const TestConfig* TestModule::FindTestConfigBydoublefieldunique(double doublefieldunique)
{
	const auto itr = m_mapTestConfigBydoublefieldunique.find(doublefieldunique);
	if (itr == m_mapTestConfigBydoublefieldunique.end())
	{
		return nullptr;
	}

	const TestConfig* ret = itr->second;
	return ret;
}




void TestModule::Clear()
{

	m_vecTestConfig.clear();


	m_mapTestConfigBystringfieldunique.clear();
	m_mapTestConfigByboolfieldunique.clear();
	m_mapTestConfigBycharfieldunique.clear();
	m_mapTestConfigByint16fieldunique.clear();
	m_mapTestConfigByintfieldunique.clear();
	m_mapTestConfigByint64fieldunique.clear();
	m_mapTestConfigByuint8fieldunique.clear();
	m_mapTestConfigByuint16fieldunique.clear();
	m_mapTestConfigByuintfieldunique.clear();
	m_mapTestConfigByuint64fieldunique.clear();
	m_mapTestConfigByfloatfieldunique.clear();
	m_mapTestConfigBydoublefieldunique.clear();


}




bool TestModule::LoadTestConfig(std::string basePath)
{
	std::string jsonPath = basePath + "TestConfig.json";
	std::string jsonText;
	if (!tool::GetFileText(jsonPath.c_str(), jsonText))
	{
		std::cout << "[error] parse failed, not found the file <" << jsonPath << ">!\n";
		return false;
	}
	
    rapidjson::Document doc;
    doc.Parse<0>(jsonText.c_str());
    if (doc.HasParseError()) {
        rapidjson::ParseErrorCode code = doc.GetParseError();
        std::cout << "[error] parse <" << jsonPath << "> failed, error code = " << code << "\n";
        return false;
    }
	
	if (!doc.IsArray())
	{
		std::cout << "[error] parse <" << jsonPath << "> failed, the rapidjson doc is not array, type = " << doc.GetType() << "\n";
		return false;
	}
	
	int n = doc.Size();
	m_vecTestConfig.resize(n);
	
	for (int i = 0; i < n; ++i)
	{
		// 逐行提取json元素
		rapidjson::Value &v = doc[i];
		
		TestConfig& row = m_vecTestConfig[i];

		row.stringfield = v["stringfield"].GetString();
		
		row.boolfield = v["boolfield"].GetInt();
		
		row.charfield = v["charfield"].GetInt();
		
		row.int16field = v["int16field"].GetInt();
		
		row.intfield = v["intfield"].GetInt();
		
		row.int64field = v["int64field"].GetInt();
		
		row.uint8field = v["uint8field"].GetInt();
		
		row.uint16field = v["uint16field"].GetInt();
		
		row.uintfield = v["uintfield"].GetInt();
		
		row.uint64field = v["uint64field"].GetInt();
		
		row.floatfield = (float)v["floatfield"].GetDouble();
		
		row.doublefield = v["doublefield"].GetDouble();
		
		row.stringfieldprimary = v["stringfieldprimary"].GetString();
		
		row.boolfieldprimary = v["boolfieldprimary"].GetInt();
		
		row.charfieldprimary = v["charfieldprimary"].GetInt();
		
		row.int16fieldprimary = v["int16fieldprimary"].GetInt();
		
		row.intfieldprimary = v["intfieldprimary"].GetInt();
		
		row.int64fieldprimary = v["int64fieldprimary"].GetInt();
		
		row.uint8fieldprimary = v["uint8fieldprimary"].GetInt();
		
		row.uint16fieldprimary = v["uint16fieldprimary"].GetInt();
		
		row.uintfieldprimary = v["uintfieldprimary"].GetInt();
		
		row.uint64fieldprimary = v["uint64fieldprimary"].GetInt();
		
		row.floatfieldprimary = (float)v["floatfieldprimary"].GetDouble();
		
		row.doublefieldprimary = v["doublefieldprimary"].GetDouble();
		
		row.stringfieldunique = v["stringfieldunique"].GetString();
		
		row.boolfieldunique = v["boolfieldunique"].GetInt();
		
		row.charfieldunique = v["charfieldunique"].GetInt();
		
		row.int16fieldunique = v["int16fieldunique"].GetInt();
		
		row.intfieldunique = v["intfieldunique"].GetInt();
		
		row.int64fieldunique = v["int64fieldunique"].GetInt();
		
		row.uint8fieldunique = v["uint8fieldunique"].GetInt();
		
		row.uint16fieldunique = v["uint16fieldunique"].GetInt();
		
		row.uintfieldunique = v["uintfieldunique"].GetInt();
		
		row.uint64fieldunique = v["uint64fieldunique"].GetInt();
		
		row.floatfieldunique = (float)v["floatfieldunique"].GetDouble();
		
		row.doublefieldunique = v["doublefieldunique"].GetDouble();
		
		tool::split(v["stringfieldarray"].GetString(), row.vecstringfieldarray);
		
		tool::split_str_to_vec<bool>(v["boolfieldarray"].GetString(), row.vecboolfieldarray, tool::str_to_bool);
		
		tool::split_str_to_vec<char>(v["charfieldarray"].GetString(), row.veccharfieldarray, tool::str_to_int);
		
		tool::split_str_to_vec<int>(v["int16fieldarray"].GetString(), row.vecint16fieldarray, tool::str_to_int);
		
		tool::split_str_to_vec<int>(v["intfieldarray"].GetString(), row.vecintfieldarray, tool::str_to_int);
		
		tool::split_str_to_vec<long>(v["int64fieldarray"].GetString(), row.vecint64fieldarray, tool::str_to_int64);
		
		tool::split_str_to_vec<int>(v["uint8fieldarray"].GetString(), row.vecuint8fieldarray, tool::str_to_int);
		
		tool::split_str_to_vec<int>(v["uint16fieldarray"].GetString(), row.vecuint16fieldarray, tool::str_to_int);
		
		tool::split_str_to_vec<int>(v["uintfieldarray"].GetString(), row.vecuintfieldarray, tool::str_to_int);
		
		tool::split_str_to_vec<int64_t>(v["uint64fieldarray"].GetString(), row.vecuint64fieldarray, tool::str_to_int64);
		
		tool::split_str_to_vec<float>(v["floatfieldarray"].GetString(), row.vecfloatfieldarray, tool::str_to_float);
		
		tool::split_str_to_vec<double>(v["doublefieldarray"].GetString(), row.vecdoublefieldarray, tool::str_to_double);
		
		tool::split_str_set(v["stringfieldset"].GetString(), row.setstringfieldset);
		
		tool::split_str_to_set<bool>(v["boolfieldset"].GetString(), row.setboolfieldset, tool::str_to_bool);
		
		tool::split_str_to_set<char>(v["charfieldset"].GetString(), row.setcharfieldset, tool::str_to_int);
		
		tool::split_str_to_set<int>(v["int16fieldset"].GetString(), row.setint16fieldset, tool::str_to_int);
		
		tool::split_str_to_set<int>(v["intfieldset"].GetString(), row.setintfieldset, tool::str_to_int);
		
		tool::split_str_to_set<long>(v["int64fieldset"].GetString(), row.setint64fieldset, tool::str_to_int64);
		
		tool::split_str_to_set<int>(v["uint8fieldset"].GetString(), row.setuint8fieldset, tool::str_to_int);
		
		tool::split_str_to_set<int>(v["uint16fieldset"].GetString(), row.setuint16fieldset, tool::str_to_int);
		
		tool::split_str_to_set<int>(v["uintfieldset"].GetString(), row.setuintfieldset, tool::str_to_int);
		
		tool::split_str_to_set<int64_t>(v["uint64fieldset"].GetString(), row.setuint64fieldset, tool::str_to_int64);
		
		tool::split_str_to_set<float>(v["floatfieldset"].GetString(), row.setfloatfieldset, tool::str_to_float);
		
		tool::split_str_to_set<double>(v["doublefieldset"].GetString(), row.setdoublefieldset, tool::str_to_double);
		
		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		
			m_mapTestConfigBystringfieldunique[row.stringfieldunique] = &row;
		

		
			m_mapTestConfigByboolfieldunique[row.boolfieldunique] = &row;
		

		
			m_mapTestConfigBycharfieldunique[row.charfieldunique] = &row;
		

		
			m_mapTestConfigByint16fieldunique[row.int16fieldunique] = &row;
		

		
			m_mapTestConfigByintfieldunique[row.intfieldunique] = &row;
		

		
			m_mapTestConfigByint64fieldunique[row.int64fieldunique] = &row;
		

		
			m_mapTestConfigByuint8fieldunique[row.uint8fieldunique] = &row;
		

		
			m_mapTestConfigByuint16fieldunique[row.uint16fieldunique] = &row;
		

		
			m_mapTestConfigByuintfieldunique[row.uintfieldunique] = &row;
		

		
			m_mapTestConfigByuint64fieldunique[row.uint64fieldunique] = &row;
		

		
			m_mapTestConfigByfloatfieldunique[row.floatfieldunique] = &row;
		

		
			m_mapTestConfigBydoublefieldunique[row.doublefieldunique] = &row;
		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

		

	}
	
	return true;
}



bool TestModule::Load(std::string basePath)
{
	Clear();
		
	if (!LoadTestConfig(basePath))
	{
		return false;
	}
	
	return true;
}










